jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ListenerHolder.java

```diff
@@ -126,9 +127,67 @@ public void doStop() throws Exception
         }
     }
 
+    private EventListener wrap(final EventListener listener)
+    {
+        EventListener ret = listener;
+        ServletContextHandler contextHandler = getServletHandler().getServletContextHandler();
+        if (contextHandler != null)
+        {
+            for (ListenerHolder.WrapperFunction wrapperFunction : contextHandler.getBeans(ListenerHolder.WrapperFunction.class))
+            {
+                ret = wrapperFunction.wrapEventListener(ret);
+            }
+        }
+        return ret;
+    }
+
+    private static EventListener unwrap(EventListener listener)
+    {
+        EventListener unwrapped = listener;
+        while (ListenerHolder.WrapperEventListener.class.isAssignableFrom(unwrapped.getClass()))
+        {
+            unwrapped = ((ListenerHolder.WrapperEventListener)unwrapped).getWrappedListener();
+        }
+        return unwrapped;
+    }
```

Comment:
It would be good to have the bulk of `wrap` and `unwrap` in BaseHolder... but I'm struggling to think of how to do this nicely... as `getBeans` needs a real type.   Maybe we could pass it that type and have a common wrapper function interface:
```java
    private T wrap(T item, Class<? extends WrapperFunction<T>> wrapperType)
￼    {
￼        ServletContextHandler contextHandler = getServletHandler().getServletContextHandler();
￼        if (contextHandler != null)
￼        {
￼            for (WrapperFunction<T> wrapperFunction : contextHandler.getBeans(wrapperType))
￼            {
￼                item = wrapperFunction.wrapEventListener(item);
￼            }
￼        }
￼        return item;
￼    }
```

Note sure if that works.... or if it is better... thoughts?

Reply:
How about ...

``` java
    protected <C,W> C wrap(final C component, final Class<W> wrapperFunctionType, final BiFunction<W,C,C> function)
    {
        C ret = component;
        ServletContextHandler contextHandler = getServletHandler().getServletContextHandler();
        if (contextHandler != null)
        {
            for (W wrapperFunction : contextHandler.getBeans(wrapperFunctionType))
            {
                ret = function.apply(wrapperFunction, ret);
            }
        }
        return ret;
    }
```

With a call like the following in FilterHolder ...

``` java
_filter = wrap(_filter, FilterHolder.WrapperFunction.class, FilterHolder.WrapperFunction::wrapFilter);
```

Reply:
it's abstract > abstract > abstract.  
not the prettiest thing.
I kinda like the local wrap methods instead (easier to follow)

Reply:
I agree that is too ugly.... let me have a play with it for 15 minutes to see if I can find something simpler....

Reply:
@joakime I've pushed an alternative